### PR TITLE
Move amm functions to library #20

### PR DESCRIPTION
**Pull request by midnight-commit**
Compiles and should be fine but untested

Added:
DexLibrary
JoeStrategyV3
DexStrategyV6

### CODE REVIEW

File: contracts/lib/DexLibrary.sol [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        return getAmountOut(amountIn, reserve0, reserve1);
+    }
+
+    /**
+     * @notice Converts reward tokens to deposit tokens
+     * @dev No price checks enforced
+     * @param amount reward tokens
+     * @return deposit tokens
+     */
+    function convertWAVAXToDepositTokens(uint amount, address depositToken, IPair swapPairToken0, IPair swapPairToken1) internal returns (uint, uint) {
```
Comment (ASlowCheetah) [inline contracts/lib/DexLibrary.sol:51]: Check `_convertRewardTokensToDepositTokens` from the DexV6, this function could be generalized like that one just passing an extra argument "ERC20 rewardToken"

File: contracts/strategies/DexStrategyV6.sol [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        require(IERC20(token).transfer(to, value), 'DexStrategyV6::TRANSFER_FROM_FAILED');
+    }
+
+    /**
+     * @notice Converts reward tokens to deposit tokens
+     * @dev No price checks enforced
+     * @param amount reward tokens
+     * @return deposit tokens
+     */
+    function _convertRewardTokensToDepositTokens(uint amount) private returns (uint) {
```
Comment (ASlowCheetah) [inline contracts/strategies/DexStrategyV6.sol:188]: then we would be able to remove this function

File: contracts/strategies/DexStrategyV6.sol [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        if (adminFee > 0) {
+            _safeTransfer(address(rewardToken), owner(), adminFee);
+        }
+
+        uint reinvestFee = amount.mul(REINVEST_REWARD_BIPS).div(BIPS_DIVISOR);
+        if (reinvestFee > 0) {
+            _safeTransfer(address(rewardToken), msg.sender, reinvestFee);
+        }
+
+        uint depositTokenAmount = _convertRewardTokensToDepositTokens(
```
Comment (ASlowCheetah) [inline contracts/strategies/DexStrategyV6.sol:156]: and call the library here

File: contracts/strategies/JoeStrategyV3.sol [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        if (adminFee > 0) {
+            _safeTransfer(address(WAVAX), owner(), adminFee);
+        }
+
+        uint reinvestFee = amount.mul(REINVEST_REWARD_BIPS).div(BIPS_DIVISOR);
+        if (reinvestFee > 0) {
+            _safeTransfer(address(WAVAX), msg.sender, reinvestFee);
+        }
+
+        (uint amountOutToken0, uint amountOutToken1) = DexLibrary.convertWAVAXToDepositTokens(
```
Comment (ASlowCheetah) [inline contracts/strategies/JoeStrategyV3.sol:211]: And here would work too, we would just need to make sure we pass the argument rewardToken as the WAVAX defined in the contract

**Review by ASlowCheetah — CHANGES_REQUESTED** [review]
One last review before we're ready to merge :D

**Comment by midnight-commit** [comment]
Very good suggestions, @ASlowCheetah, thank you! Not sure why I missed those.

One thing I noticed is that "convertRewardTokensToDepositTokens" actually does more than it says by also adding the liquidity at the end. This kind of bothers me. I think we should change the name to something like "addRewardToLiquidity" or remove "addLiquidity" from "convertRewardTokensToDepositTokens" and make it a separate call in the strategies.
What do you think?

**Comment by ASlowCheetah** [comment]
I I think this function is ok, the name says it all, it's supposed to take the reward token(any), and convert that to the deposit token(which is a liquidity token), the underlying means in which it does that is just technical details, but I do agree that it is currently doing way more than it should in one block of code, nothing to worry about right now

**Pull request merged**